src/core/unittest/RecvBufferTest.cpp

```diff
@@ -19,362 +19,614 @@
+        while (Entry != &RecvBuf.Chunks) {
+            auto Chunk = CXPLAT_CONTAINING_RECORD(Entry, QUIC_RECV_CHUNK, Link);
+            printf("  %p: Len=%u, Ext=%u", Chunk, Chunk->AllocLength, Chunk->ExternalReference);
+            if (Entry == RecvBuf.Chunks.Flink) {
+                printf(", Start=%u, Len=%u", RecvBuf.ReadStart, RecvBuf.ReadLength);
+            }
+            // Print hex output of buffer
+            for (i = 0; i < Chunk->AllocLength; ++i) {
+                if (i % 16 == 0) printf("\n    ");
+                printf("%02X ", Chunk->Buffer[i]);
+            }
+            printf("\n");
+            Entry = Entry->Flink;
+        }
+        printf("\n");
+    }
+};
+
+struct WithMode : public ::testing::TestWithParam<QUIC_RECV_BUF_MODE> {
+    static void SetUpTestSuite() {}
+    static void TearDownTestSuite() {}
 };
 
-TEST(RecvBufferTest, Alloc)
+TEST_P(WithMode, Alloc)
 {
     RecvBuffer RecvBuf;
-    ASSERT_EQ(QUIC_STATUS_SUCCESS, RecvBuf.Initialize(true));
+    ASSERT_EQ(QUIC_STATUS_SUCCESS, RecvBuf.Initialize(GetParam()));
 }
 
-TEST(RecvBufferTest, AllocWithChunk)
+TEST_P(WithMode, AllocWithChunk)
 {
     RecvBuffer RecvBuf;
-    ASSERT_EQ(QUIC_STATUS_SUCCESS, RecvBuf.Initialize(true, true));
+    ASSERT_EQ(QUIC_STATUS_SUCCESS, RecvBuf.Initialize(GetParam(), true));
 }
 
-void TestSingleWriteRead(uint16_t WriteLength, uint64_t WriteOffset, uint64_t DrainLength = 0)
+void TestSingleWriteRead(QUIC_RECV_BUF_MODE Mode, uint16_t WriteLength, uint64_t WriteOffset, uint64_t DrainLength)
 {
     RecvBuffer RecvBuf;
-    ASSERT_EQ(QUIC_STATUS_SUCCESS, RecvBuf.Initialize());
+    ASSERT_EQ(QUIC_STATUS_SUCCESS, RecvBuf.Initialize(Mode));
     uint64_t InOutWriteLength = DEF_TEST_BUFFER_LENGTH; // FC limit same as recv buffer size
-    BOOLEAN ReadyToRead = FALSE;
+    BOOLEAN NewDataReady = FALSE;
     ASSERT_EQ(
         QUIC_STATUS_SUCCESS,
         RecvBuf.Write(
             WriteOffset,
             WriteLength,
             &InOutWriteLength,
-            &ReadyToRead));
-    ASSERT_EQ(WriteOffset == 0, ReadyToRead != FALSE); // Only ready to read if we wrote to the front
+            &NewDataReady));
+    ASSERT_EQ(WriteOffset == 0, NewDataReady != FALSE); // Only ready to read if we wrote to the front
+    ASSERT_EQ(WriteOffset == 0, RecvBuf.HasUnreadData());
     ASSERT_EQ(WriteLength+WriteOffset, InOutWriteLength); // All data was newly written
     ASSERT_EQ(WriteLength+WriteOffset, RecvBuf.GetTotalLength()); // Total length should be offset plus write length
     uint64_t ReadOffset;
-    QUIC_BUFFER ReadBuffer;
-    ASSERT_EQ(
-        WriteOffset == 0, // If we didn't write to the front, then it's not ready to read yet
-        RecvBuf.Read(&ReadOffset, &ReadBuffer));
-    if (WriteOffset != 0) return; // Nothing else to validate if we didn't write to the front
+    QUIC_BUFFER ReadBuffers[3];
+    uint32_t BufferCount = ARRAYSIZE(ReadBuffers);
+    if (!NewDataReady) return; // Can't read if it's not ready
```

Comment:
Not needed? You did an assert earlier.

Reply:
No, because different modes have different behaviors.

Reply:
The assert `ASSERT_EQ(WriteOffset == 0, NewDataReady != FALSE);` indicates that it's possible for `NewDataReady` to be `FALSE`. So on this line, we bail in that scenario. Because if we were to try to read when not ready, that would assert in the product code (you're not supposed to do that).